runtime/src/accounts_db.rs

```diff
@@ -424,12 +407,13 @@ impl AccountsDB {
         let mut infos: Vec<AccountInfo> = vec![];
         while infos.len() < with_meta.len() {
             let storage = self.fork_storage(fork_id);
+            let mut count_and_status = storage.count_and_status.write().unwrap();
```

Comment:
This doesn't seem right. Seems like there is still a window between when the storage was chosen and taking the lock here.

Reply:
I think that should be fine - the critical section happens after the count_and_status lock is grabbed, which is actually appending to the AppendVec. Because the functions that check this count_and_status counter and then call reset() also first grab the lock on count_and_status, this means that these set of "reset functions" (set_status() and remove_account) and this append() function should be serialized with respect to each other, so no thread should reset the AppendVec after another thread has already appended to it.

Reply:
t1: selects a storage (with zero accounts in it), is suspended before grabbing the lock
t2: calls remove_dead_accounts(), finds the empty storage, removes it
t1: successfully appends to a removed storage

Reply:
@carllin it should set the state to “acquired” when a thread grabs it.

Reply:
Yea, that's the case I was also thinking of.

Reply:
@rwalker, @sakridge, In order for the AccountStorageEntry to be removed, cleanup_dead_fork here: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L431 makes sure the fork is less than the last known root. So the only way that race can happen is if the thread is calling add_account() on a bank that is < the current root, which shouldn't be possible?

Reply:
ok, catching up.

next issue I see is that you've serialized all store_accounts() calls, because holding the lock while we append_accounts() will cause fork_storage() will block on x.status()...

by the way, that filter_map() could check fork.id before acquiring the lock from x.status()...

Reply:
@rob-solana, that's true even now, no? All writes to the same AppendVec in `append_accounts` are serialized on the append_offset() lock. I think in order to add concurrent access, we can 

1) Do what @aeyakovenko suggested and add  a "state" to the AppendVec and set it to `acquired`, so we don't have to hold the count_and_status lock.

2) Implementing #4476 as Anatoly suggested in a comment above, which should allow for concurrent writes.

Reply:
@carllin #4523 

^^^ this eliminates anyone holding a state read/write lock while doing I/O